### DISCUSSION TITLE
HAProxy configuration as load balancer for Keycloak cluster

### DISCUSSION THREAD

**Discussion by DmitryLkk** [keycloak/keycloak#35954] [Adapter deprecation]
### Description

Hello, everyone!
I am running keycloak 26.0.6. 
I have two nodes of Keycloak cluster(it works fine). And 1 node for HAProxy(doesnt work as expected 502 error).
`ldi-keycloak-node01 172.25.48.83
ldi-keycloak-node02 172.25.48.84
ldi-haproxy-node01 172.25.48.82`

I am looking for the right configuration for my /etc/haproxy/haproxy.cfg
_my keycloak.conf for the first node(for the second the same except https-host=172.25.48.84_

db=postgres
db-username=keycloak
db-password=keycloak
db-url=jdbc:postgresql://172.25.48.88:5000/keycloak
https-certificate-file=/opt/keycloak/keycloak-26.0.6/certs/keycloak.crt.pem
https-certificate-key-file=/opt/keycloak/keycloak-26.0.6/certs/keycloak.key.pem
https-host=172.25.48.83
https-port=8080
hostname-strict=false
cache=ispn
cache-config-file=cache-ispn-tcpping.xml
health-enabled=true
metrics-enabled=true
proxy-headers=xforwarded`

cache-ispn-tcpping.xml contains the following:

    <jgroups>
        <stack name="tcpping" extends="tcp">
        <TCP bind_port="7800"/>
        <TCPPING
    initial_hosts="172.25.48.83[7800],172.25.48.84[7800]"
        port_range="0"
    stack.combine="REPLACE"
    stack.position="MPING"
        />
        </stack>
    </jgroups>
    .......(standart lines here)......

my broken HAProxy config, which I found here [https://medium.com/@omarbenabdejlil.dev/how-to-configure-keycloak-cluster-with-ispn-cache-20f323cdb4cd](url)

global
        log /dev/log    local0
        log /dev/log    local1 notice
        chroot /var/lib/haproxy
        stats socket /run/haproxy/admin.sock mode 660 level admin expose-fd listeners
        stats timeout 30s
        user haproxy
        group haproxy
        daemon

defaults
        log     global
        mode    http
        option  httplog
        option  dontlognull
        timeout connect 5000
        timeout client  50000
        timeout server  50000
        errorfile 400 /etc/haproxy/errors/400.http
        errorfile 403 /etc/haproxy/errors/403.http
        errorfile 408 /etc/haproxy/errors/408.http
        errorfile 500 /etc/haproxy/errors/500.http
        errorfile 502 /etc/haproxy/errors/502.http
        errorfile 503 /etc/haproxy/errors/503.http
        errorfile 504 /etc/haproxy/errors/504.http

frontend keycloakhttp
        bind *:80
        redirect scheme https if !{ ssl_fc }

frontend keycloakhttps
        bind *:443 ssl crt /opt/keycloak.pem # this .pem(crt + key) was generated on haproxy node. Maybe here is the mistake.
        mode http
        default_backend keycloak_nodes

backend keycloak_nodes
        mode http
        balance roundrobin
        server kc1 172.25.48.83:8080 check
        server kc2 172.25.48.84:8080 check

Hope someone helps with the correct config for HAProxy. And also how to generate certificates the right way.
Thanks in advance!

**Comment by shawkins** [comment]
Check your haproxy logs. Sometimes there are additional system setting needed for things to be reachable by haproxy.

**Comment by DmitryLkk** [comment]
Thanks for advice.
The problem was in certs. I copied certs that I generated on haproxy node to keycloak nodes. And it worked. Also I added ssl no verify in haproxy.cfg because it was the test environment.  I also used "curl https://127.0.0.1:443". "curl: (60) SSL certificate problem: self-signed certificate. curl failed to verify the legitimacy of the server and therefore could not
establish a secure connection to it."
